js/background.js

```diff
@@ -90,18 +90,37 @@
   storage.fetch();
 
   const idleDetector = new IdleDetector();
+  let isMigrationWithIndexComplete = false;
+  let isMigrationWithoutIndexComplete = false;
   idleDetector.on('idle', async () => {
     const NUM_MESSAGES_PER_BATCH = 1;
-    const database = Migrations0DatabaseWithAttachmentData.getDatabase();
-    const batch = await MessageDataMigrator.processNextBatchWithoutIndex({
-      databaseName: database.name,
-      minDatabaseVersion: database.version,
-      numMessagesPerBatch: NUM_MESSAGES_PER_BATCH,
-      upgradeMessageSchema,
-    });
-    console.log('Upgrade message schema:', batch);
 
-    if (batch.done) {
+    if (!isMigrationWithIndexComplete) {
+      const batchWithIndex = await MessageDataMigrator.processNext({
+        BackboneMessage: Whisper.Message,
+        BackboneMessageCollection: Whisper.MessageCollection,
+        numMessagesPerBatch: NUM_MESSAGES_PER_BATCH,
+        upgradeMessageSchema,
+      });
+      console.log('Upgrade message schema (with index):', batchWithIndex);
+      isMigrationWithIndexComplete = batchWithIndex.done;
+    }
+
+    if (!isMigrationWithoutIndexComplete) {
+      const database = Migrations0DatabaseWithAttachmentData.getDatabase();
+      const batchWithoutIndex = await MessageDataMigrator.processNextBatchWithoutIndex({
+        databaseName: database.name,
+        minDatabaseVersion: database.version,
+        numMessagesPerBatch: NUM_MESSAGES_PER_BATCH,
+        upgradeMessageSchema,
+      });
+      console.log('Upgrade message schema (without index):', batchWithoutIndex);
+      isMigrationWithoutIndexComplete = batchWithoutIndex.done;
+    }
```

Comment:
Running both migrations, based on just `id` index as well as more sophisticated `schemaVersion` index 👀 Could you please review and double check all the variables are correctly assigned? Thanks!

Reply:
Interesting. It looks like there's no permanent way to declare the migration _without_ index complete. I can see how we'd repeatedly want to run the migration with index, always checking to see if we've left old messages behind. But this implementation suggests that forevermore, on startup, we'll run each of these to see if any more items need processing. Off the top of my head I think I would have wanted a flag in the app to say that we're done with the migration without index, since it's a one time (though still very epic) change. For example, new installations will never need to run it at all - so that flag can be set on link.

Reply:
Great suggestion! We already have a flag but we never set it on first run. I just fixed that based on your feedback 😄 50a0048bf2b1f1e620ca820289089ae05503c4fe